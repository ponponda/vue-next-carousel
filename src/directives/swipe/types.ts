export interface SwipeOptions {
    watch: boolean; // watch moving coordinates
    onStart?: (coordinates: SwipeCoordinates) => void;
    onUp?: (coordinates: SwipeCoordinates) => void;
    onRight?: (coordinates: SwipeCoordinates) => void;
    onDown?: (coordinates: SwipeCoordinates) => void;
    onLeft?: (coordinates: SwipeCoordinates) => void;
    onEnd?: (coordinates: SwipeCoordinates) => void;
}

export type SwipeCoordinates = {
    x: number;
    y: number;
    xDiff?: number;
    yDiff?: number;
};
