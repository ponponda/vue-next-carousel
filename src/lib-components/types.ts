export const CarouselSymbol = Symbol("VCarousel");

export interface CarouselContext {
    getSelectedIndex(): number;
    getSourceLength(): number;
    getImgWidth(): number;
}
