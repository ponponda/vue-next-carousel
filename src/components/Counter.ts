import { defineComponent, h, inject } from "vue";
import { CarouselContext, CarouselSymbol } from "./types";

export default /*#__PURE__*/ defineComponent({
    name: "VCarouselCounter",
    setup() {
        const carouselCtx = inject<CarouselContext>(CarouselSymbol);

        return () =>
            carouselCtx?.getSourceLength()
                ? h(
                      "div",
                      {
                          class: "v-carousel-counter",
                      },
                      `${carouselCtx.getSelectedIndex() +
                          1}/${carouselCtx?.getSourceLength()}`
                  )
                : null;
    },
});
