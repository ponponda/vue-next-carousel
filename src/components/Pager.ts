import { defineComponent, h, inject } from "vue";
import { CarouselContext, CarouselSymbol } from "./types";

export default /*#__PURE__*/ defineComponent({
    name: "VCarouselPager",
    emits: ["click"],
    setup(_, { emit }) {
        const carouselCtx = inject<CarouselContext>(CarouselSymbol);

        return () =>
            carouselCtx?.getSourceLength()
                ? h(
                      "div",
                      { class: "v-carousel-pager" },
                      [...Array(carouselCtx?.getSourceLength()).keys()].map(
                          (idx) =>
                              h("button", {
                                  class: {
                                      "v-carousel-current-page":
                                          carouselCtx?.getSelectedIndex() ===
                                          idx,
                                  },
                                  onClick: () => emit("click", idx),
                              })
                      )
                  )
                : null;
    },
});
