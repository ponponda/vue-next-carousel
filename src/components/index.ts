import { App } from "vue";
import VCarousel from "./Carousel";

const plugin = {
    ...VCarousel,
    install(app: App) {
        app.component("VCarousel", VCarousel);
    },
};

export default plugin;
