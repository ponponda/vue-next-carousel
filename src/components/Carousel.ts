import {
    defineComponent,
    h,
    provide,
    Comment,
    VNode,
    withDirectives,
    watch,
    computed,
    reactive,
} from "vue";
import { CarouselContext, CarouselSymbol } from "./types";
import ResizeObserver from "../utils/resize-observer";
import Swipe from "../directives/swipe";
import Slide from "./Slide";
import Pager from "./Pager";
import Counter from "./Counter";

export default /*#__PURE__*/ defineComponent({
    name: "VCarousel",
    props: {
        autoplay: {
            type: Boolean,
            default: true,
        },
        autoplayTimeout: {
            type: Number,
            default: 3000,
        },
        loop: {
            type: Boolean,
            default: true,
        },
        minSwipeDistance: {
            type: Number,
            default: 100,
        },
        slideRemainRatio: {
            type: Number,
            default: 0.2,
            validator: (v: number) => v >= 0 && v < 1,
        },
        transitionDuration: {
            type: Number,
            default: 300,
        },
    },
    setup(props, { slots }) {
        const {
            element: carouselRef,
            contentRect: carouselRect,
        } = new ResizeObserver();
        const source =
            slots.default?.().filter((e) => e.type !== Comment) || [];
        const sourceLength = source.length;
        const loopSource = ([] as VNode[]).concat(source);

        const config =
            props.loop && sourceLength > 1
                ? reactive({
                      slideRemainRatio: props.slideRemainRatio,
                      startIdx: 1,
                      endIdx: sourceLength + 1,
                      touchMovement: 0,
                      selectedIdx: 1,
                      transitionDuration: props.transitionDuration,
                  })
                : reactive({
                      slideRemainRatio: props.slideRemainRatio,
                      startIdx: 0,
                      endIdx: sourceLength - 1,
                      touchMovement: 0,
                      selectedIdx: 0,
                      transitionDuration: props.transitionDuration,
                  });

        let autoplay: NodeJS.Timer | undefined;

        const windowWidth = computed(() => carouselRect.value?.width || 0);
        const blankWidth = computed(
            () => (windowWidth.value * config.slideRemainRatio) / 2
        );
        const imgWidth = computed(
            () => windowWidth.value * (1 - config.slideRemainRatio)
        );

        const slideTranslate = computed(
            () =>
                -1 * imgWidth.value * config.selectedIdx +
                blankWidth.value +
                config.touchMovement
        );

        const carouselCtx: CarouselContext = {
            getSelectedIndex: () => {
                if (props.loop) {
                    if (config.selectedIdx === 0) return sourceLength - 1;
                    else if (config.selectedIdx === config.endIdx) return 0;
                }

                return config.selectedIdx - config.startIdx;
            },
            getSourceLength: () => sourceLength,
            getImgWidth: () => imgWidth.value,
        };

        provide(CarouselSymbol, carouselCtx);

        const stopAutoplay = () => {
            if (autoplay) clearInterval(autoplay);
        };

        const slideTo = (idx: number) => {
            config.transitionDuration = props.transitionDuration;

            if (idx > config.endIdx) {
                config.selectedIdx = config.startIdx;
            } else if (idx < 0) {
                config.selectedIdx = config.endIdx;
            } else {
                config.selectedIdx = idx;
            }
        };
        const slideToNext = () => {
            slideTo(config.selectedIdx + 1);
        };
        const slideToPrev = () => {
            slideTo(config.selectedIdx - 1);
        };

        if (props.loop && sourceLength > 1) {
            loopSource.unshift(source[sourceLength - 1]);
            loopSource.push(source[0]);
        }

        if (props.autoplay && sourceLength > 1) {
            autoplay = setInterval(() => {
                slideToNext();
            }, props.autoplayTimeout);
        }

        watch(
            () => config.selectedIdx,
            (current) => {
                setTimeout(() => {
                    config.transitionDuration = 0;
                    if (props.loop) {
                        if (current === config.endIdx) {
                            config.selectedIdx = config.startIdx;
                        } else if (current === 0) {
                            config.selectedIdx = config.endIdx - 1;
                        }
                    }
                }, 200);
            }
        );

        return () =>
            h(
                "div",
                { class: "v-carousel", ref: carouselRef },
                loopSource.length > 0
                    ? [
                          withDirectives(
                              h(
                                  "div",
                                  {
                                      class: "v-carousel-slides-wrapper",
                                      style: {
                                          "transition-duration":
                                              config.transitionDuration + "ms",
                                          transform: `translate3d(${slideTranslate.value}px, 0, 0)`,
                                      },
                                  },
                                  loopSource.map((e, idx) =>
                                      h(
                                          Slide,
                                          {
                                              active:
                                                  idx === config.selectedIdx,
                                          },
                                          () => e
                                      )
                                  )
                              ),
                              [
                                  [
                                      Swipe,
                                      {
                                          watch: true,
                                          onStart: () => {
                                              config.transitionDuration = 0;
                                              stopAutoplay();
                                          },
                                          onRight: ({
                                              xDiff,
                                          }: {
                                              xDiff: number;
                                          }) => {
                                              config.touchMovement = xDiff;
                                          },
                                          onLeft: ({
                                              xDiff,
                                          }: {
                                              xDiff: number;
                                          }) => {
                                              config.touchMovement = xDiff;
                                          },
                                          onEnd: ({
                                              xDiff,
                                          }: {
                                              xDiff: number;
                                          }) => {
                                              const pass =
                                                  Math.abs(xDiff) >
                                                  props.minSwipeDistance;

                                              config.transitionDuration =
                                                  props.transitionDuration;
                                              config.touchMovement = 0;
                                              if (pass) {
                                                  if (xDiff > 0) {
                                                      slideToPrev();
                                                  } else {
                                                      slideToNext();
                                                  }
                                              }
                                          },
                                      },
                                  ],
                              ]
                          ),
                          h(Pager, {
                              onClick: (v: number) => {
                                  slideTo(v + 1);
                                  stopAutoplay();
                              },
                          }),
                          h(Counter),
                      ]
                    : []
            );
    },
});
