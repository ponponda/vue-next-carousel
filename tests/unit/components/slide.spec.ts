import { shallowMount } from "@vue/test-utils";
import { h } from "vue";
import VCarouselSlide from "../../../src/components/Slide";
import { CarouselSymbol } from "../../../src/components/types";

describe("Slide Component", () => {
    it("should match snapshot", () => {
        const wrapper = shallowMount(VCarouselSlide, {
            props: { active: true },
            global: {
                provide: {
                    [CarouselSymbol]: {
                        getImgWidth: () => 200,
                    },
                },
            },
            slots: {
                default: [h("div")],
            },
        });

        expect(wrapper.html()).toMatchSnapshot();
    });
});
