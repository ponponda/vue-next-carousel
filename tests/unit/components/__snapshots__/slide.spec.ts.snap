// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slide Component should match snapshot 1`] = `
"<div class=\\"v-carousel-slide-active\\" style=\\"flex: 0 0 200px;\\">
  <anonymous-stub></anonymous-stub>
</div>"
`;
